qutebrowser/mainwindow/prompt.py

```diff
@@ -711,6 +728,35 @@ def _init_fileview(self):
         # The model needs to be sorted so we get the correct first/last index
         self._file_model.directoryLoaded.connect(
             lambda: self._file_model.sort(0))
+    
+    def _get_valid_dirs(self, path):
+        dirs = []
+        try:
+            num_rows = self._file_model.rowCount(self._root_index)
+            for row in range(num_rows):
+                index = self._file_model.index(row, 0, self._file_model.index(path))
+                hidden = self._to_complete not in index.data()
+                if not hidden:
+                    dirs.append(index.data())
+                # self._file_view.setRowHidden(index.row(), index.parent(), hidden)
+        except FileNotFoundError:
+            log.prompt.debug("Directory doesn't exist, can't \
+                             get valid dirs")
+        
+        return dirs
+
+    def _do_completion(self, idx, which):
+        filename = self._file_model.fileName(idx)
+        valid_dirs = self._get_valid_dirs(self._current_path)
```

Comment:
There's a lot of code duplication between `_get_valid_dirs` and `_directories_hide_show_model`. Couldn't you use something like `while idx.isValid() and not self._file_view.isIndexHidden(idx):` in the loop below?

Reply:
This may just be because I am tired, but I can't get it to work without `_get_valid_dirs` therefore I will keep it in there until further notice.